### PR TITLE
WIP Add WrappedArray helper and adapt AcquiredArraysOf to use it

### PR DESCRIPTION
**Pull request by gavin-norman-sociomantic**

### CODE REVIEW

**Comment by mihails-strasuns-sociomantic** [comment]
Looks ocean'ish

**Comment by gavin-norman-sociomantic** [comment]
Yeah, I wondered that.

File: src/swarm/neo/util/AcquiredResources.d [hunk 1]
```diff
@@ -14,6 +14,113 @@ module swarm.neo.util.AcquiredResources;
 
 import ocean.transition;
 
+private struct WrappedArray ( T )
```
Comment (gavin-norman-sociomantic) [inline src/swarm/neo/util/AcquiredResources.d:17]: Note: the name of the struct is chosen pretty much at random. We can probably think of something better.

**Comment by gavin-norman-sociomantic** [comment]
There are several aspects to this PR:
1. A new feature: the wrapped array template.
2. A bug fix: using the wrapped array internally in `AcquiredArraysOf` and `Acquired`.
3. A breaking change: changing the return type of `AcquiredArraysOf.acquire`. This should probably be implemented as a deprecation (of the existing method) and a new feature (adding a new method with a different return type).

**Comment by gavin-norman-sociomantic** [comment]
1 & 2: https://github.com/sociomantic-tsunami/swarm/pull/131

File: src/swarm/neo/util/AcquiredResources.d [hunk 2]
```diff
@@ -14,6 +14,113 @@ module swarm.neo.util.AcquiredResources;
...
+        for ( size_t i = this.length - 1; i >= 0; i-- )
+        {
+            res = dg(i, *this.getElement(i));
+            if ( res )
+                break;
+        }
+        return res;
+    }
+
+    private T* getElement ( size_t t_index )
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/util/AcquiredResources.d:113]: Why both this and `opIndex`?

Comment (nemanja-boric-sociomantic) [reply src/swarm/neo/util/AcquiredResources.d:113]: I see, I guess, to remove duplication in `opIndex` and `opIndexAssign`.

Comment (nemanja-boric-sociomantic) [reply src/swarm/neo/util/AcquiredResources.d:113]: But, could we implement `opIndexAssign` in terms of `opIndex`?

**Comment by gavin-norman-sociomantic** [comment]
Breaking this PR down into smaller chunks. (#131 is the first.)

**Pull request closed without merging**